src/util/hashFetch/index.js

```diff
@@ -0,0 +1,247 @@
+const fsRmdir = util.promisify(fs.rmdir);
+
+const MAX_DOWNLOADED_FILE_SIZE_BYTES = 10485760; // 20MB
+
+function registerFailedAttemptAndThrow (hash, url, e) {
+  ExpoRetry.get().registerFailedAttempt(hash, url);
+  throw e;
+}
+
+function checkHashMatch (hash, path) {
+  return fsReadFile(path).then(content => {
+    if (sha3(content) !== `0x${hash}`) { throw new Error(`Hashes don't match: expected 0x${hash}, got ${sha3(content)}`); }
+  });
+}
+
+// List directory contents and tell if each item is a file or a directory
+function ls (folderPath) {
+  return fsReaddir(folderPath)
+    .then(filenames =>
+      Promise.all(filenames.map(filename => {
+        const filePath = path.join(folderPath, filename);
+
+        return fsStat(filePath).then(stat => ({ isDirectory: stat.isDirectory(), filePath, filename }));
+      }))
+    );
+}
+
+function rawUnzipTo (zipPath, extractPath) {
+  return new Promise((resolve, reject) => {
+    const unzipParser = unzip.Extract({ path: extractPath });
+
+    fs.createReadStream(zipPath).pipe(unzipParser);
+
+    unzipParser.on('error', reject);
+
+    unzipParser.on('close', resolve);
+  });
+}
+
+function unzipThroughTo (tempPath, extractPath, finalPath) {
+  return rawUnzipTo(tempPath, extractPath)
+    .then(() => fsUnlink(tempPath))
+    .then(() => ls(extractPath))
+    .then(files => {
+      // Check if the zip file contained a root folder
+      if (files.length === 1 && files[0].isDirectory) {
+        // Rename the root folder (contaning the dapp) to finalPath
+        const rootFolderPath = files[0].filePath;
+
+        return fsRename(rootFolderPath, finalPath)
+            .then(() => fsRmdir(extractPath));
+      } else {
+        // No root folder: extractPath contains the dapp
+        return fsRename(extractPath, finalPath);
+      }
+    });
+}
+
+function download (url, destinationPath) {
+  const file = fs.createWriteStream(destinationPath); // Will replace any existing file
+
+  return new Promise((resolve, reject) => {
+    const httpx = url.startsWith('https:') ? https : http;
+
+    httpx.get(url, response => {
+      var size = 0;
+
+      response.on('data', (data) => {
+        size += data.length;
+
+        if (size > MAX_DOWNLOADED_FILE_SIZE_BYTES) {
+          response.destroy();
+          response.unpipe(file);
+          fsUnlink(destinationPath);
+          reject(`File download aborted: exceeded maximum size of ${MAX_DOWNLOADED_FILE_SIZE_BYTES} bytes`);
+        }
+      });
+
+      response.pipe(file);
+
+      file.on('finish', () => {
+        file.close(resolve);
+      });
+    });
+  });
+}
+
+function hashDownload (hash, url, zip = false) {
+  const tempFilename = `${hash}.part`;
+  const tempPath = path.join(getHashFetchPath(), 'partial', tempFilename);
+
+  const finalPath = path.join(getHashFetchPath(), 'files', hash);
+
+  return download(url, tempPath)
+      // Don't register a failed attempt if the download failed; the user might be offline.
```

Comment:
We should actually distinguish between:
1. User offline
2. Size limit reached
3. Server returned non-200 response

We can only skip registering failed attempt in the first case, but skipping it in the other two opens up a ddos vector

Reply:
You're right. I think it's easier for now to register the failed attempt in any case, and assume the user is online.